changeset/commit.go

```diff
@@ -17,113 +17,66 @@ limitations under the License.
+	err       error
+	once      sync.Once
 
-var commitIDRE = regexp.MustCompile(`^[a-f0-9]{40}$`)
+	// for testing
+	readBuildInfo = debug.ReadBuildInfo
+)
 
-// Get tries to fetch the first 7 digitals of GitHub commit ID from HEAD file in
-// KO_DATA_PATH. If it fails, it returns the error it gets.
+// Get returns the 'vcs.revision' property from the embedded build information
+// This function will return an error if value is not a valid Git SHA
+//
+// The result will have a '-dirty' suffix if the workspace was not clean
 func Get() (string, error) {
-	data, err := readFileFromKoData(commitIDFile)
-	if err != nil {
-		return "", err
+	once.Do(func() {
+		rev, err = get()
+	})
+
+	return rev, err
+}
+
+func get() (string, error) {
+	info, ok := readBuildInfo()
+	if !ok {
+		return "", errors.New("unable to read build info")
 	}
-	commitID := strings.TrimSpace(string(data))
-	if rID := strings.TrimPrefix(commitID, "ref: "); rID != commitID {
-		// First try to read from the direct ref file - e.g. refs/heads/main
-		data, err := readFileFromKoData(rID)
-		if err != nil {
-			if !os.IsNotExist(err) {
-				return "", err
-			}
-
-			// Ref file didn't exist - it might be contained in the packed-refs
-			// file.
-			var pferr error
-			data, pferr = findPackedRef(rID)
-			// Only return the sub-error if the packed-refs file exists, otherwise
-			// just let the original error return (e.g. treat it as if we didn't
-			// even attempt the operation). This is primarily to keep the error
-			// messages clean.
-			if pferr != nil {
-				if os.IsNotExist(pferr) {
-					return "", err
-				}
-				return "", pferr
-			}
+
+	var revision string
+	var modified bool
+
+	for _, s := range info.Settings {
+		switch s.Key {
+		case "vcs.revision":
+			revision = s.Value
+		case "vcs.modified":
+			modified, _ = strconv.ParseBool(s.Value)
 		}
-		commitID = strings.TrimSpace(string(data))
 	}
-	if commitIDRE.MatchString(commitID) {
-		return commitID[:7], nil
-	}
-	return "", fmt.Errorf("%q is not a valid GitHub commit ID", commitID)
-}
 
-// readFileFromKoData tries to read data as string from the file with given name
-// under KO_DATA_PATH then returns the content as string. The file is expected
-// to be wrapped into the container from /kodata by ko. If it fails, returns
-// the error it gets.
-func readFileFromKoData(filename string) ([]byte, error) {
-	f, err := koDataFile(filename)
-	if err != nil {
-		return nil, err
+	if revision == "" {
+		return Unknown, nil
 	}
-	defer f.Close()
-	return ioutil.ReadAll(f)
-}
 
-// readFileFromKoData tries to open the file with given name under KO_DATA_PATH.
-// The file is expected to be wrapped into the container from /kodata by ko.
-// If it fails, returns the error it gets.
-func koDataFile(filename string) (*os.File, error) {
-	koDataPath := os.Getenv(koDataPathEnvName)
-	if koDataPath == "" {
-		return nil, fmt.Errorf("%q does not exist or is empty", koDataPathEnvName)
+	if shaRegexp.MatchString(revision) {
```

Comment:
Technically, we don't need this check 😎 

(not necessarily advice to remove it, just a funny coincidence)

Actually, I wonder if this should just be `if len(revision) >= 7`, so folks using Subversion or whatever can still enjoy this goodness. I'm not sure there aren't other places we assume Git in and around here, but just in case.

Reply:
I think I'd leave it as - since I don't want to assume truncating will work for different VCS - ie svn is numeric so we'd be alternating the commit id

Reply:
A further improvement could be to do this only for vcs that use SHA's for revisions -ie git, hg etc.